url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/ACS.java

```diff
@@ -0,0 +1,1521 @@
+				}
+			}
+		}
+		//Stochastically decide an action.
+		if(shortlist == null)
+		{
+			//This statement should NEVER be true, but if it does, then the ACS will return the default external 
+			//"DO_NOTHING" action.
+			ChosenAction = PossibleActions.get("DO_NOTHING_EXTERNAL");
+			shortlist = new LinkedList <AbstractAction> ();
+			shortlist.add(ChosenAction);
+		}
+		else
+		{
+			ChosenAction = (AbstractAction)SELECTOR.select(shortlist);
+		}
+		
+		if(ModuleUsed != ModuleTypes.BL)
+		{
+			for(AbstractRule r : usageCandidates)
+			{
+				if(ChosenAction.equals(r.getAction()))
+				{
+					r.addTimeStamp(CurrentTimeStamp);
+				}
+			}
+		}
+		
+		for(AbstractImplicitModule i : BLImplicitModuleStore)
+		{
+			AbstractOutputChunk o = i.getOutput(ChosenAction.getID());
+			if(o != null)
+				i.setChosenOutput(o);
+		}
+		return ChosenAction;
+	}
+	
+	/**
+	 * Selects a rule type to use on the top level when performing action decision making with stochastic
+	 * level selection. If the top level is not to be used, this method returns null.
+	 * @return The rule collection to use on the top level or null if the top level is not to be used.
+	 */
+	private RuleCollection selectType ()
+	{
+		double pBL;
+		double pRER;
+		double pIRL;
+		double pFR;
+		if(LEVEL_SELECTION_OPTION == LevelSelectionOptions.FIXED)
+		{
+			pBL = FIXED_BL_PROBABILITY * (MCS_A + (MCS_B * MCS_BL_PROBABILITY));
+			pRER = FIXED_RER_PROBABILITY;
+			pIRL = FIXED_IRL_PROBABILITY;
+			pFR = FIXED_FR_PROBABILITY;
+		}
+		else
+		{
+			double srBL = VARIABLE_BL_BETA * ((VARIABLE_C3 + BLImplicitModuleStore.PM)/(VARIABLE_C4 + BLImplicitModuleStore.PM + BLImplicitModuleStore.NM));
+			double srRER = VARIABLE_RER_BETA * ((VARIABLE_C3 + RERRuleStore.PM)/(VARIABLE_C4 + RERRuleStore.PM + RERRuleStore.NM));
+			double srIRL = VARIABLE_IRL_BETA * ((VARIABLE_C3 + IRLRuleStore.PM)/(VARIABLE_C4 + IRLRuleStore.PM + IRLRuleStore.NM));
+			double srFR = VARIABLE_FR_BETA * ((VARIABLE_C3 + FRRuleStore.PM)/(VARIABLE_C4 + FRRuleStore.PM + FRRuleStore.NM));
+			pBL = srBL/(srBL + srRER + srIRL + srFR);
+			pBL = pBL * (MCS_A + (MCS_B * MCS_BL_PROBABILITY));
+			pRER = srRER/(srBL + srRER + srIRL + srFR);
+			pIRL = srIRL/(srBL + srRER + srIRL + srFR);
+			pFR = srFR/(srBL + srRER + srIRL + srFR);
+		}
+		double normalizer = pBL + pRER + pIRL + pFR;
+		pBL = pBL/normalizer;
+		pRER = pRER/normalizer;
+		pIRL = pIRL/normalizer;
+		pFR = pFR/normalizer;
+		double p = Math.random();
+		double pp = 1;
+		pp -= pRER;
+		if(p >= pp)
+		{
+			ModuleUsed = ModuleTypes.RER;
+			return RERRuleStore;
+		}
+		pp -= pIRL;
+		if(p >= pp)
+		{
+			ModuleUsed = ModuleTypes.IRL;
+			return IRLRuleStore;
+		}
+		pp -= pFR;
+		if(p >= pp)
+		{
+			ModuleUsed = ModuleTypes.FR;
+			return FRRuleStore;
+		}
+		return null;
+	}
+	
+	/**
+	 * Combines the outcomes of the top and bottom levels when level combination is being used.
+	 */
+	private void performCombination (LinkedList <AbstractRule> usageCandidates)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 21. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.